### PR TITLE
[RSDK-2519] add version information

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-2519)

Equivalent RPlidar PR: https://github.com/viamrobotics/rplidar/pull/36

Moved from: https://github.com/viamrobotics/viam-cartographer/pull/73

1. log model, git sha & tag information as first log line (same as [viam-server](https://github.com/viamrobotics/rdk/blob/main/web/server/entrypoint.go#L77))
2. terminate the program early if `--version` or `-version` is provided via argv

Example logs:
```
# CLI
➜  viam-cartographer git:(RSDK-2519-add-version-information) ./bin/cartographer-module -version
1.683750429744546e+09   info    cartographerModule      module/main.go:45       viam:slam:cartographer  {"git_rev": "60a1cdeb9a5461e3e9b4a24de9e0e5b69001e252"}
➜  viam-cartographer git:(RSDK-2519-add-version-information) ./bin/cartographer-module --version
1.68375043212148e+09    info    cartographerModule      module/main.go:45       viam:slam:cartographer  {"git_rev": "60a1cdeb9a5461e3e9b4a24de9e0e5b69001e252"}
➜  viam-cartographer git:(RSDK-2519-add-version-information) ./bin/cartographer-module --versio
1.6837504347693539e+09  info    cartographerModule      module/main.go:45       viam:slam:cartographer  {"git_rev": "60a1cdeb9a5461e3e9b4a24de9e0e5b69001e252"}
1.683750434769696e+09   info    cartographerModule      module/module.go:185    server listening at --versio
^C1.683750437859991e+09 info    cartographerModule      module/module.go:199    Shutting down gracefully.
➜  viam-cartographer git:(RSDK-2519-add-version-information) ./bin/cartographer-module not right
1.683750441910502e+09   fatal   cartographerModule      utils@v0.1.20-0.20230424163529-ce35a14fc60f/runtime.go:78       Usage: ./bin/cartographer-module [socket path] start module
or: ./bin/cartographer-module --version        print version & exit
➜  viam-cartographer git:(RSDK-2519-add-version-information) ./bin/cartographer-module
1.683750446820477e+09   fatal   cartographerModule      utils@v0.1.20-0.20230424163529-ce35a14fc60f/runtime.go:78       Usage: ./bin/cartographer-module [socket path] start module
or: ./bin/cartographer-module --version        print version & exit


# LOGS
nickpi@nick-slam-dev-pi:~ $ viam-server -config /etc/viam.json -debug
2023-05-08T16:10:34.060-0400    INFO    robot_server    server/entrypoint.go:85 Viam RDK  {"version": "v0.2.43", "git_rev": "3f0b23713482146b7dc7e5843d49cdf5abb5e4dd"}
2023-05-08T16:10:34.087-0400    DEBUG   robot_server    config/reader.go:174    reading configuration from the cloud
2023-05-08T16:10:34.313-0400    DEBUG   robot_server    rpc/dial.go:189 connected via gRPC{"address": "app.viam.com:443"}
2023-05-08T16:10:34.344-0400    DEBUG   robot_server    rpc/dial.go:189 connected via gRPC{"address": "app.viam.com:443"}
2023-05-08T16:10:34.469-0400    DEBUG   robot_server    config/reader.go:230    reading tlsCertificate from the cloud
2023-05-08T16:10:34.616-0400    DEBUG   robot_server    rpc/dial.go:189 connected via gRPC{"address": "app.viam.com:443"}
2023-05-08T16:10:34.837-0400    DEBUG   robot_server    rpc/dial.go:189 connected via gRPC{"address": "app.viam.com:443"}
2023-05-08T16:10:34.845-0400    DEBUG   robot_server    web/web.go:431  module server listening    {"socket path": "/tmp/viam-module-471300900/parent.sock"}
2023-05-08T16:10:44.146-0400    INFO    robot_server.modmanager.process.rplidar_module_/usr/local/bin/rplidar-module.StdOut        pexec/managed_process.go:224
\_ 1.6835766441466134e+09       info    rplidarModule   module/module.go:187    server listening at /tmp/viam-module-471300900/rplidar_module.sock
2023-05-08T16:10:54.974-0400    INFO    robot_server.modmanager.process.carto_module_/usr/local/bin/cartographer-module.StdOut     pexec/managed_process.go:224
\_ 1.6835766549747365e+09       info    cartographerModule      module/main.go:73       viam:slam:cartographer     {"git_rev": "566d185f18de62a7ef710ecafb10ce3e54e9fa40"}
2023-05-08T16:10:54.992-0400    INFO    robot_server.modmanager.process.carto_module_/usr/local/bin/cartographer-module.StdOut     pexec/managed_process.go:224

```

Sources: 
1. https://www.digitalocean.com/community/tutorials/using-ldflags-to-set-version-information-for-go-applications
2. https://github.com/viamrobotics/rdk/blob/main/Makefile#L9

Manual Testing:
1. Compiled & run on M1 Mac
2. Compiled in canon on M1 mac & executed on amd64 Pi.

### CODE REVIEW

File: module/main.go [hunk 1]
```diff
@@ -12,11 +12,30 @@ import (
...
 
+// Versioning variables which are replaced by LD flags.
+var (
+	Version     = "development"
+	GitRevision = ""
+)
+
+// Arguments for the command.
+type Arguments struct {
+	Version bool `flag:"version,usage=print version"`
```
Comment (EshaMaharishi) [inline module/main.go:23]: [note to self] this is using a go struct tag (https://www.digitalocean.com/community/tutorials/how-to-use-struct-tags-in-go).

in particular, the "flag" tag is indicating that if `--version` is passed on the command line, then the "Version" bool should be set to true. the "usage" tag is indicating that if `--help` (or equivalent) is passed on the command line, then the description for Version should be "print version"

**Review by EshaMaharishi — APPROVED** [review]
LGTM! Thanks for getting this improvement in!

File: Makefile [hunk 2]
```diff
@@ -1,6 +1,9 @@
 BUILD_CHANNEL?=local
 TOOL_BIN = bin/gotools/$(shell uname -s)-$(shell uname -m)
 PATH_WITH_TOOLS="`pwd`/$(TOOL_BIN):${PATH}"
+GIT_REVISION = $(shell git rev-parse HEAD | tr -d '\n')
+TAG_VERSION?=$(shell etc/tag_version.sh)
+LDFLAGSOPTS = -ldflags "-X 'main.Version=${TAG_VERSION}' -X 'main.GitRevision=${GIT_REVISION}'"
```
Comment (nicksanford) [inline Makefile:6]: I changed this from LDFLAGS as CMAKE was acting on this ENV var during C++ compilation & failing to link carto  C++ code.

File: Makefile [hunk 3]
```diff
@@ -1,6 +1,9 @@
 BUILD_CHANNEL?=local
 TOOL_BIN = bin/gotools/$(shell uname -s)-$(shell uname -m)
 PATH_WITH_TOOLS="`pwd`/$(TOOL_BIN):${PATH}"
+GIT_REVISION = $(shell git rev-parse HEAD | tr -d '\n')
+TAG_VERSION?=$(shell git tag --points-at | sort -Vr | head -n1)
```
Comment (nicksanford) [inline Makefile:5]: This follows james's advice here: https://github.com/viamrobotics/rplidar/pull/36#discussion_r1189060802

**Review by Otterverse — APPROVED** [review]
LGTM

File: Makefile (hunk 3 above)
Comment (jeremyrhyde) [reply Makefile:5]: validated against RDK

File: module/main.go [hunk 4]
```diff
@@ -12,11 +14,42 @@ import (
...
+
 func main() {
 	utils.ContextualMain(mainWithArgs, golog.NewLogger("cartographerModule"))
 }
 
 func mainWithArgs(ctx context.Context, args []string, logger golog.Logger) error {
+	var versionFields []interface{}
+	// print version on statup
+	// exit 0 if --version or -version argv is provided
+	// NOTE: utils.ParseFlags can't be used w/o a nolint as it
```
Comment (jeremyrhyde) [inline module/main.go:31]: update comment to not discuss nolint that no longer exists

File: module/main.go [hunk 5]
```diff
@@ -12,11 +14,42 @@ import (
...
+	// errors if there are args not described by the struct,
+	// which is the case during normal module use.
+	if len(args) != 2 {
```
Comment (jeremyrhyde) [inline module/main.go:34]: Why is this length 2? If you run `carto-module --version` isn't that just 1 arg?

**Review by jeremyrhyde — CHANGES_REQUESTED** [review]
Question regarding number of args

File: module/main.go (hunk 5 above)
Comment (Otterverse) [reply module/main.go:34]: In linux/unix, ARGV[0] is the name of the command itself. Go does this is well with os.Args (which is where this ultimately comes from.

File: module/main.go (hunk 4 above)
Comment (nicksanford) [reply module/main.go:31]: done

**Pull request merged**